hi-ml-multimodal/src/health_multimodal/image/model/encoder.py

```diff
@@ -0,0 +1,173 @@
+        """Image encoder forward pass."""
+
+        patch_emb = self.encoder(current_image)
+        avg_pooled_emb = torch.flatten(torch.nn.functional.adaptive_avg_pool2d(patch_emb, (1, 1)), 1)
+        if return_patch_embeddings:
+            return patch_emb, avg_pooled_emb
+
+        return avg_pooled_emb
+
+    def reload_encoder_with_dilation(self, replace_stride_with_dilation: Optional[Sequence[bool]] = None) -> None:
+        """Workaround for enabling dilated convolutions after model initialization.
+
+        :param replace_stride_with_dilation: for each layer to replace the 2x2 stride with a dilated convolution
+        """
+        if self.img_model_type == "resnet18":
+            # resnet18 uses BasicBlock implementation, which does not support dilated convolutions.
+            raise NotImplementedError("resnet18 does not support dilated convolutions")
+
+        if replace_stride_with_dilation is None:
+            replace_stride_with_dilation = False, False, True
+
+        device = next(self.encoder.parameters()).device
+        new_encoder = self._create_encoder(replace_stride_with_dilation=replace_stride_with_dilation).to(device)
+
+        if self.encoder.training:
+            new_encoder.train()
+        else:
+            new_encoder.eval()
+
+        new_encoder.load_state_dict(self.encoder.state_dict())
+        self.encoder = new_encoder
+
+
+class MultiImageEncoder(ImageEncoder):
+    """Multi-image encoder trunk module for the ``ImageModel`` class.
+    It can be used to encode multiple images into combined latent representation.
+    """
+
+    def __init__(self, img_model_type: str):
+        super().__init__(img_model_type)
+
+        output_dim = 256  # The aggregate feature dim of the encoder is `2 * output_dim` i.e. [f_static, f_diff]
+        grid_shape = (14, 14)  # Spatial dimensions of patch grid.
+
+        backbone_output_feature_dim = get_encoder_output_dim(self.encoder, device=get_module_device(self))
+
+        self.backbone_to_vit = nn.Conv2d(in_channels=backbone_output_feature_dim, out_channels=output_dim,
+                                         kernel_size=1, stride=1, padding=0, bias=False)
+        self.vit_pooler = VisionTransformerPooler(input_dim=output_dim, grid_shape=grid_shape)
+
+        # Missing image embedding
+        self.missing_previous_emb = nn.Parameter(torch.zeros(1, output_dim, 1, 1))
+        trunc_normal_(self.missing_previous_emb, std=.02)
+
+    def forward(self,  # type: ignore[override]
+                current_image: torch.Tensor,
+                previous_image: Optional[torch.Tensor] = None,
+                return_patch_embeddings: bool = False) -> TypeImageEncoder:
+
+        batch_size = current_image.shape[0]
+
+        if previous_image is not None:
+            assert current_image.shape == previous_image.shape
+            x = super().forward(torch.cat([current_image, previous_image], dim=0), return_patch_embeddings=True)[0]
```

Comment:
Could we split this up into multiple lines? Quite a bit happening here.

Reply:
Yes, I agree. Updated.